File: opcua-webservice-spring/core/src/main/java/org/opcfoundation/webserver/addressspace/nodes/UaVariable.java
package org.opcfoundation.webserver.addressspace.nodes;

import org.eclipse.milo.opcua.sdk.core.AccessLevel;
import org.eclipse.milo.opcua.stack.core.AttributeId;
import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.eclipse.milo.opcua.stack.core.types.builtin.NodeId;
import org.eclipse.milo.opcua.stack.core.types.builtin.Variant;
import org.eclipse.milo.opcua.stack.core.types.builtin.unsigned.UByte;
import org.eclipse.milo.opcua.stack.core.types.enumerated.IdType;
import org.eclipse.milo.opcua.stack.core.types.enumerated.NodeClass;
import org.springframework.lang.Nullable;
import org.opcfoundation.webserver.addressspace.nodes.builtin.UaReferenceTypes;

import java.util.List;
import java.util.Map;

public class UaVariable extends UaInstanceNode {
    private final UaVariableType typeDefinition;
    private final NodeId dataType;
    private int valueRank;
    private int accessLevel;
    private boolean historizing;
    private Variant value;

    public UaVariable(
            NodeId nodeId,
            String browseName,
            LocalizedText displayName,
            NodeId dataType,
            int valueRank,
            int accessLevel,
            UaVariableType typeDefinition)
    {
        super(nodeId, browseName, displayName);

        this.typeDefinition = typeDefinition;
        this.addReference(new UaReference(typeDefinition, UaReferenceTypes.HasTypeDefinition, true));

        this.dataType = dataType;
        this.valueRank = valueRank;
        this.accessLevel = accessLevel;
        this.value = Variant.NULL_VALUE;
    }

    @Override
    public NodeClass nodeClass() {
        return NodeClass.Variable;
    }

    public UaVariableType typeDefinition() {
        return typeDefinition;
    }

    public NodeId dataType() {
        return dataType;
    }

    public int valueRank() {
        return valueRank;
    }

    public void setValueRank(int valueRank) {
        this.valueRank = valueRank;
    }

    public int accessLevel() {
        return accessLevel;
    }

    public void setAccessLevel(int accessLevel) {
        this.accessLevel = accessLevel;
    }

    public boolean historizing() { return historizing; }

    public void setHistorizing(boolean historizing) { this.historizing = historizing; }

    public Variant value() {
        return value;
    }

    public void setValue(Variant value)
    {
        this.value = value;
    }

    public void addMember(UaVariable member)
    {
        addMemberNode(member);
    }

    public @Nullable UaVariable addMember(String name)
    {
        UaInstanceNode member = typeDefinition.getMember(name);
        if (null == member || NodeClass.Variable != member.nodeClass()) return null;

        UaVariable memberVariable = (UaVariable)member;

        String newMemberIdValue = nodeId().getIdentifier().toString();
        newMemberIdValue += "-";
        newMemberIdValue += memberVariable.browseName();

        UaVariable newVariable = new UaVariable(
                new NodeId(nodeId().getNamespaceIndex(),newMemberIdValue),
                memberVariable.browseName(),
                memberVariable.displayName(),
                memberVariable.dataType(),
                memberVariable.valueRank(),
                AccessLevel.CurrentRead.getValue(),
                memberVariable.typeDefinition());

        addMemberNode(newVariable);

        return newVariable;
    }

    @Override
    public Variant getAttribute(int attributeId)
    {
        if (AttributeId.Value.id() == attributeId)
        {
            return value;
        } else if (AttributeId.DataType.id() == attributeId) {
            return new Variant(dataType);
        } else if (AttributeId.ValueRank.id() == attributeId) {
            return new Variant(valueRank);
        } else if (AttributeId.AccessLevel.id() == attributeId) {
            return new Variant(UByte.valueOf(accessLevel));
        } else if (AttributeId.UserAccessLevel.id() == attributeId) {
            return new Variant(UByte.valueOf(accessLevel));
        } else if (AttributeId.MinimumSamplingInterval.id() == attributeId) {
            return new Variant((double)0);
        } else if (AttributeId.Historizing.id() == attributeId) {
            return new Variant(false);
        }

        return super.getAttribute(attributeId);
    }
}

File: opcua-webservice-node/src/server/addressspace/nodes/UaVariable.ts
import { Attributes, NodeClass } from "opcua-webapi";
import { UaAccessLevel, UaLocalizedText, UaNodeId, UaVariant, UaVariantType, VariableTypeIds } from "opcua-webapi-ts";
import { UaInstanceNode, UaReference, UaReferenceTypes, UaVariableType } from ".";

export class UaVariable extends UaInstanceNode {
    private readonly _typeDefinition: UaVariableType;
    private readonly _dataType: UaNodeId;
    private _valueRank: number;
    private _accessLevel: number;
    private _historizing: boolean;
    private _value: UaVariant;

    constructor(
        nodeId: UaNodeId,
        browseName: string,
        displayName: UaLocalizedText,
        dataType: UaNodeId,
        valueRank: number,
        accessLevel: number,
        typeDefinition: UaVariableType
    ) {
        super(nodeId, browseName, displayName);
        this._typeDefinition = typeDefinition;
        this.addReference(new UaReference(typeDefinition, UaReferenceTypes.HasTypeDefinition, true));
        this._dataType = dataType;
        this._valueRank = valueRank;
        this._accessLevel = accessLevel;
        this._historizing = false;
        this._value = UaVariant.null();
    }

    get nodeClass(): NodeClass {
        return NodeClass.Variable;
    }

    get typeDefinition(): UaVariableType {
        return this._typeDefinition;
    }

    get dataType(): UaNodeId {
        return this._dataType;
    }

    get valueRank(): number {
        return this._valueRank;
    }

    set valueRank(valueRank: number) {
        this._valueRank = valueRank;
    }

    get accessLevel(): number {
        return this._accessLevel;
    }

    set accessLevel(accessLevel: number) {
        this._accessLevel = accessLevel;
    }

    get historizing(): boolean {
        return this._historizing;
    }

    set historizing(historizing: boolean) {
        this._historizing = historizing;
    }

    get value(): UaVariant {
        return this._value;
    }

    set value(value: UaVariant) {
        this._value = value;
    }

    addMember(member: UaVariable)
    {
        this.addMemberNode(member);
    }

    addMemberByName(name: string): UaVariable | null {
        const member = this.typeDefinition.getMember(name);
        if (member === null || member.nodeClass !== NodeClass.Variable) {
            return null;
        }

        const memberVariable = member as UaVariable;
        const newMemberIdValue = `${this.nodeId.value.toString()}-${memberVariable.browseName}`;
        const newVariable = new UaVariable(
            new UaNodeId(newMemberIdValue, this.nodeId.nsIndex),
            memberVariable.browseName,
            memberVariable.displayName,
            memberVariable.dataType,
            memberVariable.valueRank,
            UaAccessLevel.CurrentRead,
            memberVariable.typeDefinition
        );

        this.addMemberNode(newVariable);
        return newVariable;
    }

    getAttribute(attributeId: number): UaVariant {
        if (Attributes.Value === attributeId) {
            return this._value;
        }
        if (Attributes.DataType === attributeId) {
            return UaVariant.nodeId(this._dataType);
        }
        if (Attributes.ValueRank === attributeId) {
            return UaVariant.integer(this._valueRank, UaVariantType.Int32);
        }
        if (Attributes.AccessLevel === attributeId) {
            return UaVariant.integer(this._accessLevel, UaVariantType.Byte);
        }
        if (Attributes.UserAccessLevel === attributeId) {
            return UaVariant.integer(this._accessLevel, UaVariantType.Byte);
        }
        if (Attributes.MinimumSamplingInterval === attributeId) {
            return UaVariant.double(0);
        }
        if (Attributes.Historizing === attributeId) {
            return UaVariant.boolean(this._historizing);
        }

        return super.getAttribute(attributeId);
    }
}